RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.h

```diff
@@ -0,0 +1,131 @@
+#ifndef CAHitQuadrupletGenerator_H
+#define CAHitQuadrupletGenerator_H
+
+#include "RecoPixelVertexing/PixelTriplets/interface/HitQuadrupletGenerator.h"
+#include "RecoTracker/TkSeedingLayers/interface/SeedComparitorFactory.h"
+#include "RecoTracker/TkSeedingLayers/interface/SeedComparitor.h"
+#include "RecoPixelVertexing/PixelTrackFitting/src/RZLine.h"
+#include "RecoTracker/TkSeedGenerator/interface/FastCircleFit.h"
+#include "RecoTracker/TkMSParametrization/interface/PixelRecoUtilities.h"
+#include "RecoTracker/TkMSParametrization/interface/LongitudinalBendingCorrection.h"
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+
+
+#include "RecoTracker/TkHitPairs/interface/HitPairGeneratorFromLayerPair.h"
+#include "RecoTracker/TkHitPairs/interface/LayerHitMapCache.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+
+class TrackingRegion;
+class HitQuadrupletGeneratorFromTripletAndLayers;
+class SeedingLayerSetsHits;
+
+namespace edm {
+    class Event;
+}
+namespace edm {
+    class EventSetup;
+}
+
+class CAHitQuadrupletGenerator : public HitQuadrupletGenerator {
+public:
+    typedef LayerHitMapCache LayerCacheType;
+
+public:
+
+    CAHitQuadrupletGenerator(const edm::ParameterSet& cfg, edm::ConsumesCollector& iC);
+
+    virtual ~CAHitQuadrupletGenerator();
+
+    /// from base class
+    virtual void hitQuadruplets(const TrackingRegion& reg, OrderedHitSeeds & triplets,
+            const edm::Event & ev, const edm::EventSetup& es);
+
+    void findQuadruplets(const TrackingRegion& region, OrderedHitSeeds& result,
+            const edm::Event& ev, const edm::EventSetup& es,
+            const SeedingLayerSetsHits::SeedingLayerSet& fourLayers);
+    
+    
+private:
+    edm::EDGetTokenT<SeedingLayerSetsHits> theSeedingLayerToken;
+
+    LayerCacheType theLayerCache;
+
+    std::unique_ptr<SeedComparitor> theComparitor;
+
+    class QuantityDependsPtEval {
```

Comment:
Maybe we should abstract `QuantityDependsPtEval` out of being inner class of `CAHitQuadrupletGenerator` and `PixelQuadrupletGenerator` instead of copy-paste (for me it's fine to cleanup in a later PR).